### ISSUE TITLE
User defined types should be parameterizable with builtin types

### ISSUE THREAD

**Issue by danabr** [j14159/mlfe#20]
Consider this example:
````
module user_defined_types

type proplist 'k 'v = list ('k, 'v)

type optlist 'v = proplist atom 'v
````

This fails with:
````erlang
{error, {badmatch, {error, {6,mlfe_parser, ["syntax error before: ",["\"atom\""]]}}}
````

The parser expects a user defined type to only take type variables. See [poly_type](https://github.com/j14159/mlfe/blob/master/src/mlfe_parser.yrl#L102) in mlfe_paser.yrl.

mlfe_typer also has this assumption.

**Comment by j14159** [comment]
Ah yes - looks like the parser only permits the base polymorphic types to do this.  Should be a pretty simple fix with error checking/validation almost entirely occurring in `mlfe_typer.erl`.  Nice find, thanks!

**Comment by j14159** [comment]
Oh, actually, maybe we just need `poly_type` to take a list of type expressions instead of type vars.

**Comment by danabr** [comment]
I also think we should not define `base_type`, `base_list`, etc in the lexer. It prevents us to use these for identifiers: This code does not compile for example:

````
map f l = 
    match l with
        [] -> []
    | h :: t ->
        (f h) :: (map f t)
````

**Comment by j14159** [comment]
Sure, I just didn't want them to be replaceable as types.

**Comment by j14159** [comment]
@danabr let me know if you'd like to take a crack at this - if not I'll start looking at it this weekend.  Just don't want to duplicate effort :)

**Comment by danabr** [comment]
I'm playing with it now. I'm not sure I will be able to make anything mergable, though. I'll let you know tomorrow.

**Comment by j14159** [comment]
No worries.  The typer is...convoluted at the moment, happy to help out.  I'll leave this issue alone in favour of working through records some more (no branch yet, just reading and note-taking).

**Comment by danabr** [comment]
I don't think I'm going to make it. I can get the parser to work somewhat as I want, but I cannot wrap my head around the typer. See [this branch] (https://github.com/j14159/mlfe/compare/master...danabr:fix/user-defined-types-parameterized-with-builtin-types) for a failing test case.

**Comment by danabr** [comment]
Seems like the issues I ran into were caused by the issue described in #23. After applying only the parser changes on top of #24 my original example seems to work. I'll try to get around to cleaning it up and creating a PR.

**Comment by danabr** [comment]
Closed by #27.
